core/src/main/resources/jenkins/model/BuildDiscarderProperty/help.html

```diff
@@ -1,19 +1,51 @@
-  Jenkins offers two criteria:
-
+  This determines when, if ever, build records for this project should be
+  discarded. Build records include the console output, archived artifacts, and
+  any other metadata related to a particular build.
+  <p>
+  Keeping fewer builds means less disk space will be used in the <i>Build Record
+  Root Directory</i>, which is specified on the <i>Configure System</i> screen.
+  <p>
+  <!-- TODO: Most/all of the following should be dynamic, showing info from the
+       `BuildDiscarder` implementations that are currently loaded -->
+  Jenkins offers two options for determining when builds should be discarded:
   <ol>
     <li>
-      Driven by age. You can have Jenkins delete a record if it reaches a certain age
-      (for example, 7 days old.)
+      Build age: discard builds if they reach a certain age; for example, seven
+      days old.
     <li>
-      Driven by number. You can have Jenkins make sure that it only maintains up to
-      N build records. If a new build is started, the oldest record will
-      be simply removed.
+      Build count: discard the oldest build if a certain number of builds
+      already exist.
   </ol>
+  These two options can be active at the same time, so you can keep builds for
+  14 days, but only up to a limit of 50 builds, for example.  If either limit is
+  exceeded, then any builds beyond that limit will be discarded.
+  <p>
+  You can also ensure that important builds are kept forever, regardless of the
+  setting here &mdash; click the <i>Keep this build forever</i> button on the
+  build page.
+  <br>
+  The last stable and last successful build are also excluded from these rules.
+
+  <hr>
+
+  In the <i>Advanced</i> section, the same options can be specified, but
+  specifically for build <b>artifacts</b>.  If enabled, build artifacts will be
+  discarded for any builds which exceed the defined limits.  The builds
+  themselves will still be kept; only the associated artifacts, if any, will be
+  deleted.
+  <p>
+  For example, if a project builds some software and produces a large installer,
+  which is archived, you may wish to always keep the console log and information
+  about which source control commit was built.  While for disk space reasons,
+  you may want to keep only the last three installers that were built.
+  <br>
+  This can make sense for projects where you can easily recreate the same
+  artifacts later by building the same source control commit again.
+
+  <hr>
 
-  Jenkins also allows you to mark an individual build as 'Keep this log forever', to
-  exclude certain important builds from being discarded automatically.
-  The last stable and last successful build are always kept as well.
-</div>
\ No newline at end of file
+  For time-based limits, note that Jenkins does not discard items as soon as
+  the configured time is exceeded; these rules are evaluated the next time that
+  a build of this project starts.
```

Comment:
*All rules* are evaluated only when the next build *finishes*. This is relevant e.g. when you reduce the number in the config and wonder why Jenkins does not start deleting builds.